### PR TITLE
code-review: update requirements

### PR DESCRIPTION
**Pull request by OleksiyRudenko**
## Major changes:

* add intro to General Requirements

* update Code Formatting section

* add note on variable/function naming (verbs, nouns etc)

* add explanation re Event Delegation

* add requirements on TJSW pre-OOP

* update requirements on OO JS (Frogger)

* update requirements on OOP Exercise

## Next steps

Once this PR finalized and merged:

1. Publish updated docs to kottans.org/documentation.
2. Update [homeworks 2022 checklists](https://github.com/kottans/frontend-2022-homeworks/tree/main/.github/pr-checklists) so that bot's messages on TJSW, Frogger and OOP Exercise lead to kottans.org/documentation sections rather than contain checklists.
3. Announce changes in Students' chat.

### CODE REVIEW

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:16]: `and to contribute comfortably`?

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:37]: `express` through `expressive`. Maybe `express the intent through descriptive variable names` is better?

Comment (artem-trubin) [reply docs/doc/code-review.md:37]: `into a properly named function` or `into properly named functions`

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:283]: `CSS`, uppercase

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:297]: `a pattern` - `it should be eliminated`

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:298]: No dot in the end.

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:311]: `are`?

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:331]: But `update` method shouldn't render anything. It is for update the state of the object which is then used for rendering in another method. Maybe a bad example?

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:344]: Better wording:
`Implement base classes for child classes to inherit shared properties from`.

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:321]: We have to figure out do we really need to restrict `class` usage for students, when some mentors are for it.

File: docs/doc/code-review.md
Comment (artem-trubin) [inline docs/doc/code-review.md:9]: Maybe remove `habr` links? Russian site, doesn't work without VPN.

Comment (OleksiyRudenko) [reply docs/doc/code-review.md:9]: Makes sense as we have links to original posts anyway and code reviewers are likely have sufficient command of English.
Cannot remember who suggested adding translations and why. Would expect yet another opinion on this matter.

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:16]: Suggesting to add a preposition and an excplicit reference to the object of contribution, for the sake of clarity.
```suggestion
The code is written for other people to read and to contribute to the code comfortably.
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:37]:
```suggestion
- Unnecessary comments should be avoided. Find a way to express the intent through descriptive variable names or by abstracting part of the code into properly named function. Comments, if there are any, should explain "why", not "what".
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:283]:
```suggestion
So, replace such parts with CSS classes. You can use `classList` to manipulate them.
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:297]:
```suggestion
1. Code is [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), which means that whenever you see a pattern in your code it should be eliminated. Examples:
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:298]:
```suggestion
   1. `print(dog); print(cat); etc ... ` should be refactored employing `Array.forEach` as the least.
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:320]:
```suggestion
1. Requirements regarding **OOP**:
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [reply docs/doc/code-review.md:321]: I'd cosider "doesn't make sense" rather than "against".
Q: Why `prototype.*` is still here? A: Legacy. The exisiting codebase is all `prototype.*`. Former Udacity theoretical part was about `prototype.*` only.

The progress in standards should be accounted for. 
I'd change this for the next course iteration. This will require to rework the original project (both codebase and documentation). Also `prototype.*` allows to tweak game engine re-defining its internals in a different file (e.g. `app.js`), which is not really possible with `class` syntax. Although no one used this opportunity yet anyway.

In a nutshell, I'd postpone this change at least until current Stage 0 end. It will also need some thoughtful effort at the expense of reviews or other improvements.

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:331]:
```suggestion
      (e.g. `update` method does only character coordinates update and
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:332]:
```suggestion
      doesn't contain any **inline** code to check e.g. collisions; calling other methods from `update` is legitimate)
```

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:344]:
```suggestion
   1. Implement base classes for child classes to inherit shared properties from.
```

**Comment by OleksiyRudenko** [comment]
@artem-trubin thank you for your review. Please review the code suggestions above.
I will merge those once we are aligned to reduce the review scope to the outstanding items.

There are two outstanding items:
- links to habr
- `prototype.*` vs `class`

Let's go on with these after smaller changes get committed.

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [reply docs/doc/code-review.md:37]: I like `descriptive` more as it more specific than `proper`.

**Review by OleksiyRudenko — COMMENTED** [review]
The suggestions below
* remove habr translation links
* revert `prototype` requirement on Frogger

**Review by suchov — APPROVED** [review]

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:53]:
```suggestion
_(although some conventional shorthands are OK; e.g. `i`, `j` to denote array indexes, `x`, `y` to denote coordinates and other similar cases)_
```

Comment (OleksiyRudenko) [inline docs/doc/code-review.md:53]:
```suggestion
_(although some conventional shorthands are OK; e.g. `i` to when looping through array by index, `x`/`y` to denote coordinates and other similar cases)_
```

**Review by OleksiyRudenko — COMMENTED** [review]
Minor update on variable naming to avoid naming like `axisX` or `valueY` in submitted code.

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:35]:
```suggestion
- Unnecessary comments should be avoided. Find a way to express the intent through descriptive variable names or by abstracting parts of the code into properly named functions. Comments, if there are any, should explain "why", not "what".
```

**Review by OleksiyRudenko — COMMENTED** [review]
Missed a comment re function**s**

File: docs/doc/code-review.md
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:359]:
```suggestion
   1. when `this.constructor.name` is used properly there is no need to specify `species` property explicitly
```

**Comment by OleksiyRudenko** [comment]
Added all suggestions.
Removed requirement on `prototype` syntax for Frogger.

**Pull request merged**
